modules/promotion/src/Element/CouponRedemptionForm.php

```diff
@@ -0,0 +1,225 @@
+<?php
+
+namespace Drupal\commerce_promotion\Element;
+
+use Drupal\commerce\Element\CommerceElementBase;
+use Drupal\commerce_order\Entity\OrderInterface;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * Provides a form element for embedding the coupon redemption form.
+ *
+ * Usage example:
+ * @code
+ * $form['coupons'] = [
+ *   '#type' => 'commerce_promotion_coupon_redemption_form',
+ *   // An order entity that valid redeemed coupons will be applied to.
+ *   '#order' => $order,
+ *   // Set to FALSE to not use AJAX submission
+ *   '#use_ajax' => FALSE,
+ * ];
+ * @endcode
+ *
+ * @RenderElement("commerce_promotion_coupon_redemption_form")
+ */
+class CouponRedemptionForm extends CommerceElementBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getInfo() {
+    $class = get_class($this);
+    return [
+      // The order which redeemed coupons will apply to.
+      '#order' => NULL,
+      '#process' => [
+        [$class, 'attachElementSubmit'],
+        [$class, 'processForm'],
+      ],
+      '#element_validate' => [
+        [$class, 'validateElementSubmit'],
+        [$class, 'validateForm'],
+      ],
+      '#commerce_element_submit' => [
+        [$class, 'submitForm'],
+      ],
+      '#theme_wrappers' => ['container'],
+    ];
+  }
+
+  /**
+   * Builds the coupon redemption form.
+   *
+   * @param array $element
+   *   The form element being processed.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   * @param array $complete_form
+   *   The complete form structure.
+   *
+   * @throws \InvalidArgumentException
+   *   Thrown when the #operation or #default_value properties are empty, or
+   *   when the #default_value property is not a valid entity.
+   *
+   * @return array
+   *   The processed form element.
+   */
+  public static function processForm(array $element, FormStateInterface $form_state, array &$complete_form) {
+    if (!$element['#order'] instanceof OrderInterface) {
```

Comment:
We can throw an \InvalidArgumentException here, makes more sense than exiting silently.